docs/datastores.md

```diff
@@ -51,6 +51,26 @@ Uses [badger](https://github.com/dgraph-io/badger) as a key value store.
 }
 ```
 
+## badger2ds
+
+Uses [badger2](https://github.com/dgraph-io/badger) as a key value store.
+
+* `syncWrites`: Flush every write to disk before continuing. Setting this to false is safe as go-ipfs will automatically flush writes to disk before and after performing critical operations like pinning. However, you can set this to true to be extra-safe (at the cost of a 2-3x slowdown when adding files).
+* `truncate`: Truncate the DB if a partially written sector is found (defaults to true). There is no good reason to set this to false unless you want to manually recover partially written (and unpinned) blocks if go-ipfs crashes half-way through a adding a file.
+* `compression`: Configure compression. When compression is enabled, every block is compressed using the specified algorithm. This option doesn't affect existing tables. Only the newly created tables are compressed. Compression can be configured to use the snappy algorithm or ZSTD with level 1, 2 or 3.
+* `blockCacheSize`: Specifies how much data cache should hold in memory. If compression is disabled, adding a cache leads to unnecessary overhead which may affect read performance.
+
+```json
+{
+	"type": "badger2ds",
+	"path": "<location of badger2 inside repo>",
+	"syncWrites": true|false,
+	"truncate": true|false,
+	"compression": "none"|"snappy"|"zstd1"|"zstd2"|"zstd3",
+	"blockCacheSize": <size of block cache>
```

Comment:
WDYT about having some value for each of these signifying the default (e.g. in https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#types we have some default values). This would allow us (the implementers) to change the default compression and blockCacheSize without users needing to be change their configs. Of course, if they wanted to fix these parameters they could explicitly put them in the config file.

My thinking is to use `null or missing` to mean "apply the default value" as we do with Flag and Priority and then define what our current default values are (no compression, no blockCacheSize).

Reply:
I would prefer that anytime a string parameter is missing or has a null (or empty `''"`) value, that this means use the default of our choosing (which could change with updates) -- _when appropriate_.  There are some cases where a value cannot change with an upgrade, such as a filesystem location. So, for these we cannot use a default for a missing or null. In those cases, we must supply a default value at initialization time, and report and error if the value is ever unspecified.

So, I do not think we can make a general statement that all missing/null string values get set to some default. This needs to be done on a case-by-case basis.  Some strings can be missing/null and a reasonable default will be used, and for others an initial value is provided but a value must always be present.

So, to address this, I have amended the description specifically for the new badger2 config values, that omitting the values from the configuration results in choosing the default setting.